### ISSUE TITLE
Update One Column in Entity issue

### ISSUE THREAD

**Issue by atak946** [abpframework/abp#3621]
Hi everyone,

i'm using `nameSpace.table.update` method but i have problem.

im updating one column but all column's updating to null,

**Example Table Columns:**

- Id
- Name
- Surname
- BornDate
- Status

**Example Data:**

```
Id = 1, Name = Murat, Surname = X, BornDate = 1995/26/06, Status = 1

nameSpace.table.update(1, { surname: "ATAK" }).done(r=>{console.log(r);});
```

**After Update Data Result**

```
Id = 1, Name = NULL, Surname = ATAK, BornDate = NULL, Status = NULL
```
why update function set NULL to all columns?

I just want to update one column, how can I do this?

Best Regards

**Comment by liangshiw** [comment]
Hi @atak946 
You should write partially updated code in your business code. This has nothing to do with abp.

**Comment by atak946** [comment]
> Hi @atak946
> You should write partially updated code in your business code. This has nothing to do with abp.
@liangshiw 

So when is the nameSpace.update function used?

I did not understand the meaning of this function

when update relational table, this function setted null all relationed id columns but i sent it all columns with filled relation id. 

**Add Columns to example table:**

- ExampleId
- public virtual tblExample ...

**My Data Before Request:**


```
Id : 1, 
Name: "Murat", 
Surname : "X", 
BornDate :"1995/06/26", 
Status : 1, 
ExampleId = 123, 
tblExample = {FILLED WITH PROPERTIES}
```

**Example Request Data:**

```
var requestData = {id : 1, name: "Murat", surname : "ATAK", bornDate :"1995/06/26", status : 1, exampleId : 123};

nameSpace.table.update(1, requestData);

```
**Result of data:**

```
Id : 1, 
Name: "Murat", 
Surname : "ATAK", 
BornDate :"1995/06/26",
Status : 1, 
ExampleId = NULL, 
tblExample = NULL
```

**Comment by liangshiw** [comment]
You should check the code of your application service update method. 
If you use the curd application service base class, you can override the base class method.

**Comment by atak946** [comment]
@liangshiw 
yes i can override but i need a general fix. the override solution is the last way for us.
thank you.

@maliming are you have any idea?

**Comment by liangshiw** [comment]
I don’t think ABP needs to change.I tested your use case and it has no problems.

Entities:
```csharp
public class Qa : Entity<int>
{
     public string Name { get; set; }

     public int? QaItemId { get; set; }
        
     public QaItem QaItem { get; set; }
}

public class QaItem: Entity<int>
{
     public string Name { get; set; }
}
```

Application Services:
```csharp
public class TestAppService : CrudAppService<Qa,QaDto, int, PagedAndSortedResultRequestDto, QaDto, QaDto>, ITestAppService
{
   public TestAppService(IRepository<Qa, int> repository) : base(repository)
   {
   }
 }
```

And test result:

![image](https://user-images.githubusercontent.com/16813853/79357429-97eb4900-7f72-11ea-8ddf-36bc4173f83f.png)

**Comment by atak946** [comment]
@liangshiw 
yes, I got a successful result on swagger, but I still have the same problem via javascript function.
can you try this?

**Comment by liangshiw** [comment]
@atak946 
![image](https://user-images.githubusercontent.com/16813853/79460786-e4449080-8027-11ea-9956-b86b630665d1.png)

**Comment by atak946** [comment]
@liangshiw 

I get successful results with fiddler, but my relational columns with javascript are null. you can see the examples in the images.

I use the same request model in fiddler and javascript

I need to understand the difference between the two I can't solve the problem

The records that I mentioned relationally are available in the tables. I checked them all and got them from the tables

**Javascript Test: JSON Data**
![JSONDATA](https://user-images.githubusercontent.com/16645386/79543897-d67a2400-8096-11ea-8798-9d68aed73a04.png)

**Javascript Test: Update Result:**
![RESPONSE](https://user-images.githubusercontent.com/16645386/79543980-ef82d500-8096-11ea-8c70-654980fcd9cf.png)

**Fiddler Test: Json Data:**
![FİDDLER](https://user-images.githubusercontent.com/16645386/79543996-f90c3d00-8096-11ea-8337-a1dcaca65d4c.png)

**Fiddler Test: Update Result:**
![FİDDLER RESPONSE](https://user-images.githubusercontent.com/16645386/79544018-03c6d200-8097-11ea-889c-8d178f5014a9.png)

```
var jsondata = {
  "equipmentNo": "string",
  "meterSerialNo": "string",
  "meterTypeId": "374e5de5-e80d-e630-d24b-39f3e0144161",
  "meterBrand": "string",
  "meterStyle": "string",
  "productionDate": "2020-04-16T08:37:52.445Z",
  "brandDate": "2020-04-16T08:37:52.445Z",
  "voltage": 0,
  "amperage": 0,
  "impulse": 0,
  "location": "string",
  "totalIndex": 0,
  "t1": 0,
  "t2": 0,
  "t3": 0,
  "t4": 0,
  "demand": 0,
  "positiveReactiveInductive": 0,
  "negativeReactiveCapacitive": 0,
  "pR_TotalIndex": 0,
  "pR_T1": 0,
  "pR_T2": 0,
  "pR_T3": 0,
  "pR_T4": 0,
  "procedurePhraseId": "1f8db326-1b47-455e-82bc-9deb98f2b8bd",
  "subscriberTitle": "string",
  "adress": "string",
  "meterRevellingDate": "2020-04-16T08:37:52.445Z",
  "meterRevellingReasonId": "28ed5b10-26b5-4cfe-ba1a-b38a937d5de7",
  "dismantledStaff": "string",
  "reportSerial": "string",
  "reportSerialNo": "string",
  "readType": 1,
  "isCompleted": true,
  "isActive": true,
  "isDeleted": true,
  "workOrderDetailId": "15c7e957-30c4-dacb-8974-39f3fa4044bb",
  "tblMeterType": null,
  "tblProcedurePhrase": null,
  "tblMeterRevellingReason": null,
  "tblWorkOrderDetails": null
};

sayacAyarServisi.workOrder.update(Model.Id, jsondata).done(function (r) {
    console.log(r);
    abp.notify.success("Akış başarıyla tamamlandı ve kapatıldı");
}).fail(function () {
    abp.notify.error("Birşeyler ters gitti lütfen daha sonra tekrar deneyin.");
});
```

**Comment by liangshiw** [comment]
Please use the CLI to reproduce, and then give me the repository

**Comment by atak946** [comment]
@liangshiw

 i'll upload tonight,

As I tested with jquery ajax, only 1 column is going to be null, as most of the problem is solved. When I try with different browsers, only 1 property in the mozilla goes null and there is a problem like chrome in the opera.

I have to identify the problem of the remaining 1 property being null.

```
$.ajax({
    url: "/api/app/workOrder/338addbf-76ea-dc48-5d8e-39f400498119",
    data: JSON.stringify(jsondata),
    contentType: "application/json",
    dataType: "json",
    type: "PUT",
    success: (r) => {
      console.log(r);
    }
});
```

![JQUERY AJAX](https://user-images.githubusercontent.com/16645386/79559663-cf5f1000-80ae-11ea-9e84-2f16a92241df.png)

**Comment by atak946** [comment]
@liangshiw 

not stable :(

![aaa](https://user-images.githubusercontent.com/16645386/79561161-89577b80-80b1-11ea-8ede-34f522ab9468.gif)

**Comment by liangshiw** [comment]
Can you give me the repository?

**Comment by atak946** [comment]
> Can you give me the repository?

https://github.com/atak946/xrepo

**Comment by liangshiw** [comment]
@atak946 I mean you use CLI to create a simple project to reproduce the issue.

**Comment by cotur** [comment]
@atak946 

> 
> 
> @liangshiw
> 
> not stable :(
> 
> ![aaa](https://user-images.githubusercontent.com/16645386/79561161-89577b80-80b1-11ea-8ede-34f522ab9468.gif)


Hi @atak946,

With that example, you are using the `Update` but in your `CustomAdvancedCrud` there is a method that named `UpdateRow` and you wrote a different bussines logic in that code but you don't use it. Maybe you need to just use `UpdateRow` method for your own `CustomAdvancedCrud` or override the `Update`.

Also in your code you did not implement the `UpdateRow` in your `ICustomAdvancedCrud` 

Also I could not run your repository, `DbMigrator` is also crashed.

So basically it is not related to ABP. The `CrudAppService` and `AutoMagically` javascript endpoints are solid stable. The problem is in your code. I suggest you to run in debug mode and watch everyline what is going on? You'll caught the bug.

**Comment by atak946** [comment]
> @atak946
> 
> > @liangshiw
> > not stable :(
> > ![aaa](https://user-images.githubusercontent.com/16645386/79561161-89577b80-80b1-11ea-8ede-34f522ab9468.gif)
> 
> Hi @atak946,
> 
> With that example, you are using the `Update` but in your `CustomAdvancedCrud` there is a method that named `UpdateRow` and you wrote a different bussines logic in that code but you don't use it. Maybe you need to just use `UpdateRow` method for your own `CustomAdvancedCrud` or override the `Update`.
> 
> Also in your code you did not implement the `UpdateRow` in your `ICustomAdvancedCrud`
> 
> Also I could not run your repository, `DbMigrator` is also crashed.
> 
> So basically it is not related to ABP. The `CrudAppService` and `AutoMagically` javascript endpoints are solid stable. The problem is in your code. I suggest you to run in debug mode and watch everyline what is going on? You'll caught the bug.

No, updaterow is only different test method not maun method. i already have this error.

**Comment by atak946** [comment]
> @atak946
> 
> > @liangshiw
> > not stable :(
> > ![aaa](https://user-images.githubusercontent.com/16645386/79561161-89577b80-80b1-11ea-8ede-34f522ab9468.gif)
> 
> Hi @atak946,
> 
> With that example, you are using the `Update` but in your `CustomAdvancedCrud` there is a method that named `UpdateRow` and you wrote a different bussines logic in that code but you don't use it. Maybe you need to just use `UpdateRow` method for your own `CustomAdvancedCrud` or override the `Update`.
> 
> Also in your code you did not implement the `UpdateRow` in your `ICustomAdvancedCrud`
> 
> Also I could not run your repository, `DbMigrator` is also crashed.
> 
> So basically it is not related to ABP. The `CrudAppService` and `AutoMagically` javascript endpoints are solid stable. The problem is in your code. I suggest you to run in debug mode and watch everyline what is going on? You'll caught the bug.

İngilizcem yetmiyor artık Ahmet bey o updaterow metodunu da denemek için ekledim sorunla uzaktan yakından ilgisi yok bu sorun 2.0 versiyonundan 2.5 versiyonuna yükseltme yaptıktan sonra çıktı.

CustomCrudAppService ile de ilgisi yok onu kaldırınca da sorun devam ediyor. Müsait olamadim müsait olunca cli ile oluşturulmuş boş projeyi de yükleyeceğim bu konuyu anlamadan kapatmanız da yanlış olmuş.

**Comment by cotur** [comment]
@atak946 So I reopened it to inspect more. Please inform me when you create the test project with cli.

**Comment by cotur** [comment]
Hi @atak946,

I've created an example application to explain detailed.

> Id = 1, Name = Murat, Surname = X, BornDate = 1995/26/06, Status = 1
> 
> nameSpace.table.update(1, { surname: "ATAK" }).done(r=>{console.log(r);});
> ```
> 
> **After Update Data Result**
> 
> ```
> Id = 1, Name = NULL, Surname = ATAK, BornDate = NULL, Status = NULL
> ```
> 
> why update function set NULL to all columns?
> 
> I just want to update one column, how can I do this?

You specified your problem here, you want to update one property of an entity, but when you use `CrudAppService` it will update only what you sent and removes others. It is about the `Mapping`

I'll give you solution step by step. 

I've created app and I have an entity that named `Contact`

As you said, when I send some data to update it will removes others because there is no good mapping in my project yet.
![resim](https://user-images.githubusercontent.com/20099739/79620398-6ad89b00-8118-11ea-8c38-4d4a4723ba7d.png)


It is occurs because the `UpdateAsync` do mapping from the input data to the orginal entity. If you do not change AutoMapper settings, it will maps what you sent.

So let's change some settings.

![resim](https://user-images.githubusercontent.com/20099739/79620214-d5d5a200-8117-11ea-9197-1048fdc35b28.png)

I just added that code to the AutoMapper settings. It will ignores the null variables. 

So the result is.

![resim](https://user-images.githubusercontent.com/20099739/79620461-a4a9a180-8118-11ea-8c82-9a2cfce51b0a.png)


It looks like the problem is fixed! 

# BUT IMPORTANT! 
It is occurs a problem, "What will happen when you want to update a property to `NULL`", so it is gonna be impossible with previous solution.
So you should look at the `AutoMapper` settings. You'll find the best solition for your implementation.

As I said before, it is not related to the ABP Framework.

**Comment by atak946** [comment]
Hi, @cotur 

Thanks to @liangshiw I understood the map problem you mentioned and I tried to send all the properties.

`I want to draw your attention to one more point, this problem happens if there is more than one relation id in the table (i have 4 column). I noticed that I did not experience this problem in other tables in my project.`

but this time I came across another problem, I tried to convey and share visuals but I guess it is a problem I cannot explain.

this is not a normal problem, why should it succeed when it first executes the same request and fails when it triggers again later. It's hard to explain.

**My problem is that although I send this field full, it will be saved as empty.**

Please read this comment.
https://github.com/abpframework/abp/issues/3621#issuecomment-615094577

after readed the comment, please look at the gif image (bottom of the next comment) 

Thank you for your interest.

**Comment by cotur** [comment]
I tried this issue at repository that shared with us by you. 

The problem is **`wrong DTO implementation`**. 

Experimental Solution:
![Atak-Problem-Solved](https://user-images.githubusercontent.com/20099739/79639117-3d303800-8192-11ea-8456-248ee5c96ea3.gif)

The `Update_X_Dto` should contains the properties what we want to update. But your `Update_X_Dto` contains several `Navigation Properties`. 

When you call `UpdateAsync` method, the `Navigation Properties` are creating with null then it occurs wrong mapping. While the result is creating, the `AutoMapper` should override the IDs' and you don't see those in your response.

As a result, we should follow `Best Practises`. :)
